File: 0006-ZigZag-Conversion/lc0006.cc
// C++ solution.

class Solution {
public:
  string convert(string s, int numRows) {
    if (numRows == 1) return s;
    vector<string> ret(numRows, "");
    int rowCursor = 0;
    bool trigger = false;

    for (char &c : s) {
      ret[rowCursor].push_back(c);
      rowCursor += trigger? -1 : 1;
      if (rowCursor == 0 || rowCursor == numRows-1) {
        trigger = !trigger;
      }
    }

    string ret_s = "";
    for (const string &piece : ret) ret_s += piece;
    return ret_s;
  }
};

File: 0006-ZigZag-Conversion/lc0006.js
// JavaScript solution.

/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */

var convert = function(s, numRows) {
    if (numRows === 1) return s;
    let ret = new Array(numRows).fill("");
    let rowCursor = 0;
    let trigger = false;

    for (c of [...s]) {
        ret[rowCursor] += c;
        rowCursor += trigger? -1 : 1;
        if (rowCursor === 0 || rowCursor === numRows-1) {
            trigger = !trigger;
        }
    }

    return ret.join('');
};
